contracts/transfer/circuits/src/gadgets/schnorr.rs

```diff
@@ -0,0 +1,206 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_plonk::constraint_system::ecc::scalar_mul::fixed_base::scalar_mul as fixed_base_scalar_mul;
+use dusk_plonk::constraint_system::ecc::scalar_mul::variable_base::variable_base_scalar_mul;
+use dusk_plonk::constraint_system::ecc::Point as PlonkPoint;
+use dusk_plonk::jubjub::{
+    AffinePoint, ExtendedPoint, GENERATOR_EXTENDED, GENERATOR_NUMS_EXTENDED,
+};
+use dusk_plonk::prelude::*;
+use plonk_gadgets::AllocatedScalar;
+use poseidon252::sponge::sponge::sponge_hash_gadget;
+use std::convert::TryInto;
+
+/// Utilises a Schnorr signature scheme,
+/// to prove the knowledge of a discrete
+/// log for a given public key.
+#[allow(non_snake_case)]
+pub fn schnorr_one_key(
+    composer: &mut StandardComposer,
+    signature: AllocatedScalar,
+    R: PlonkPoint,
+    pk: PlonkPoint,
+    message: AllocatedScalar,
+) {
+    let h = sponge_hash_gadget(composer, &[message.var]);
+    let c = sponge_hash_gadget(composer, &[*R.x(), *R.y(), h]);
+    let b = BlsScalar::zero();
+    let b = composer.add_witness_to_circuit_description(b);
+
+    let challenge = composer.xor_gate(c, b, 250);
+
+    let sig =
+        fixed_base_scalar_mul(composer, signature.var, GENERATOR_EXTENDED);
+    let p = variable_base_scalar_mul(composer, challenge, pk);
+
+    let add = sig.point().add(composer, *p.point());
+    composer.assert_equal_point(add, R);
```

Comment:
If this is a public point, the constraint should be added outside.

Reply:
Nothing is public According to Tog.